### DISCUSSION TITLE
.hpp file ending mandatory?

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#1760] [Q&A] [answered]
Hi all,

I ran into much trouble trying to compile a codebase that I already had to compile on STM32duino (using sloeber - but that is not part of the problem) some weeks (or was it already months?) ago.
Project got laid aside, I wanted to resume just to find out that I get weird error messages from the compiler.
After some fiddling around I could nail it down on a header file time.h that holds a small, simple class with a method to calculate system running time and convert it to a human-readable format.
Compiler complains that the type "class" is not known.
I the updated to latest 2.3.0 (was on 2.2.0) and the behaviour persists.
The only thing that works is renaming the header file to time.hpp
I have tested this only in a small example project that has nothing else than empty setup() and loop() as well as the #include time.h(pp)
Tested on sloeber and standard Arduino IDE.

[rant]
Since I am on windows, testing is very, very time consuming since every rebuilt after changing only smallest bits leads to a complete rebuild (thankfully not in sloeber!) which takes hilarious amount of time (5-10 minutes)
[/rant]

As to now I am quite sure that all it takes really is to rename the file to .hpp

This is quite strange, since I am almost 100% sure I changed nothing to my setup in the time between "codebase compiles" and "codebase annoys compiler and loads of errors are thrown"

**Comment by fpistm** [comment]
Hi @MasteringTheMess 
There is no time.h in the core (except one in OpenAMP fw but I guess you do not build for MP1 with serialVirtIO enabled), I guess it is the one from the arm none eabi toolchain? Anyway, I didn't found any "class" keyword and are pure C file.
We never seen this issue and I don't think rename a file is a good solution. 
I build daily on Windows and Linux. You can also refers to the GH actions and saw no errors happened. 

Maybe you do not used this core but this one ? https://github.com/rogerclarkmelbourne/Arduino_STM32

**Comment by PerennialNovice** [comment]
Hello @fpistm,

sorry for not being precise, time.h is part of my own codebase.

**Comment by fpistm** [comment]
OK so what did you expect? 
As said time.h is standard library name provided with arm none eabi toolchain. I advise to rename it to avoid any mistake: mytime.h

**Comment by PerennialNovice** [comment]
What I expect?
Some help in understanding what is going on... and hopefully a way to resolve my problem

well, in the past my Time.h got included correctly, now the compiler complains about the "class" keyword.
I didn't think that this is a problem with ambiguous includes

**Comment by fpistm** [comment, marked as answer]
You should go to the forum for this kind of issue and I answered you. Using same name than standard include is not good practice which leads to strange issue. Why it works before I can't tell moreover without any code. About subject: .hpp ending mandatory. Answer is no. 😉

**Comment by PerennialNovice** [comment]
Thank you for the advice to go to the forum!

I thought this was an issue of the stm32duino, that's why I first came here. I remember changing my time.h from struct to class because the compiler complained about something the AVR compiler gracefully ignored...
In a way the STM32 compiler helps me write better code :)

**Comment by PerennialNovice** [comment]
I just keep thinking about this...

Is there a way of finding out what filenames are "taken"?
Or is it visible in the compiler output somewhere?

**Comment by fpistm** [comment]
My guess is that time.h from the standard C library is simply not included but your one is used instead but when standard time.h is required it is probably to build c file with gcc. But as your time.h used class keyword it's failed. I don't understand why you simply not change the name of your file like this include will be done properly... This would be the same if you create stdio.h, math.h, string.h,... 
https://www.tutorialspoint.com/c_standard_library/time_h.htm

**Comment by PerennialNovice** [comment]
Of course I did this, my question is to be able to avoid such a problem in the future...
Maybe there exists a convention, that every library file (be it standard C, ARM core, STM32 core) is starts with lower case, so if I use upper case, I will always be safe?

**Reply by fpistm** [reply]
> is to be able to avoid such a problem in the future...

Conventions is simply to not use same name used by the standard C library. Using uppercase would work on Linux which is case sensitive but not on Windows (your case). On Windows: time.h == Time.H

**Comment by PerennialNovice** [comment]
maybe move this whole topic to a discussion?

**Comment by PerennialNovice** [comment]
> Conventions is simply to not use same name used by the standard C library

So I am supposed to know all these names?
What about Arduino-specific file names?
I really think it would be nice to have a little warning that the filename I chose is a bad idea

and thanks for the hint that Windows does not care about case - I tend to forget...

**Reply by fpistm** [reply]
> I really think it would be nice to have a little warning that the filename I chose is a bad idea

It is called build error 😜

**Comment by PerennialNovice** [comment]
nice one!

but if I am correct, then for some reason on several occasions it passed through without error or warning.
Not very reliable...
